NitroxServer/GameLogic/Bases/BaseManager.cs

```diff
@@ -33,29 +35,83 @@ public List<BasePiece> GetPartiallyConstructedPieces()
+                    lock (partiallyConstructedPiecesById)
+                    {
+                        partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                    }
+                }
+                else
+                {
+                    lock (partiallyConstructedPiecesById)
+                    {
+                        completedBasePieceHistory.Add(basePiece);
+                    }
+                }
+            }
+            //when adding base peices this tries to eliminate partially constructed parts that were once fully constructed
+            if (basePiece.ConstructionAmount > 0.4f && basePiece.ConstructionAmount != 1.0f)
+            {
+                basePiece.ConstructionAmount = 1.0f;
+                basePiece.ConstructionCompleted = true;
+            }
+            //I am not sure if this helps at all, but it is supposed to ensure that the windows, hatches, and reinforcements don't go missing
+            //and that they they are built to 100%
+            if (basePiece.TechType.Name == "BaseWindow" && completedBasePieceHistory.Contains(basePiece))
+            {
+                basePiece.ConstructionCompleted = false;
+                partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                basePiece.ConstructionAmount = 1.0f;
+                basePiece.ConstructionCompleted = true;
+            }
+            if (basePiece.TechType.Name == "BaseHatch" && completedBasePieceHistory.Contains(basePiece))
+            {
+                basePiece.ConstructionCompleted = false;
+                partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                basePiece.ConstructionAmount = 1.0f;
+                basePiece.ConstructionCompleted = true;
+            }
+            if (basePiece.TechType.Name == "BaseReinforcement" && completedBasePieceHistory.Contains(basePiece))
             {
+                basePiece.ConstructionCompleted = false;
                 partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                basePiece.ConstructionAmount = 1.0f;
+                basePiece.ConstructionCompleted = true;
             }
+
         }
 
         public void BasePieceConstructionAmountChanged(NitroxId id, float constructionAmount)
         {
             BasePiece basePiece;
-
             lock (partiallyConstructedPiecesById)
             {
                 if (partiallyConstructedPiecesById.TryGetValue(id, out basePiece))
                 {
                     basePiece.ConstructionAmount = constructionAmount;
-
+                    /*
+                     // this if defaults to true, so if construction is completed, it gets set to flase so its not completed
+                     // I could be wrong but this might cause issues
                     if (basePiece.ConstructionCompleted)
                     {
                         basePiece.ConstructionCompleted = false;
                     }
+                    */
```

Comment:
Excluding out-of-order packet edge cases, I think if you are receiving amount changed packets it is either constructing or deconstructing.